content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,114 @@
+
+### Added More Test Cases to Clang-Doc test suite
+
+
+Clang-doc previously had test’s which did not test the full scope of the tool especially with regards to the HTML or Markdown output of the tool. In order to make sure the optimization experiments were not accidentally breaking the tool, it was necessary to add more end to end tests. 
+In summary I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 
+
+
+### Improve clang-doc’s performance by 1.58 times 
+
+Internally the way clang-doc works is by leveraging libtooling's ASTVisitor class to parse the declaration in each project and serializing it into an internal format, which gets deserialized later when we output the final format. The bottleneck in clang-doc lied in the way clang-doc was doing redundant work when it was visiting each declaration. 
+
+For example if we had a the following project: 
+
+```cpp
+//File: Base.cpp
+
+class Base {}
+```
+
+
+```cpp
+//File: A.cpp
+
+class A : public Base {}
+```
+
+```cpp
+//File: B.cpp
+
+class B : public Base {}
+```
+
+In this case the ASTVisitor invoke by clang-doc would visit serialized the Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 
+
+The optimization ended up being a simple memoization dictionary which kept track of a list of declaration that clang-doc had visited. 
+
+Here is a plot of the benchmarking numbers:
+
+<div style="margin:0 auto;">
+  <img src="/img/clang-doc-benchmark-numbers.png"><br/>
+</div>
+
+
+### Added Template Mustache HTML Backend
+
+Clang-doc originally used an ad-hoc method of generating HTML, in this project I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however upon further discussion it was decided that the complexity of designing and implementing a new templating language was too much. 
+A LLVM community member suggested I implement a mustache as templating language. 
+Mustache was the ideal templating language, since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 
```

Comment:
```suggestion
Mustache was the ideal choice since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 
```

The bit about pushback is a bit misleading. There was some resistance, but the community member who raised the objection also wasn't aware that we already generate HTML in a number of places, and the new templating language could be used to simplify other cases as well. I think its important to put these things into context if you're going to talk about them. It's also important to chose words that make it clear we respect our colleagues opinions, even when we disagree. I don't think you've done anything to convey otherwise so far, but since I'm asking for this to be rephrased, I want to emphasize that point.